dkt/runner.go

```diff
@@ -0,0 +1,207 @@
+			fmt.Fprintf(stderr, "%v", err)
+
+			return 1
+		}
+	}
+
+	var singleArg string
+	if len(args) == 1 {
+		singleArg = args[0]
+	}
+	switch singleArg {
+	case "-v", "--version", "version":
+		// Add the runner's version before the dkt and docker-compose versions.
+		printRunnerBuildInfo("", stdout)
+	}
+
+	if debugTraceEnabled {
+		fmt.Fprintf(stderr, "%sbuilding %s\n", debugPrefix, actualMainPkg)
+	}
+	dktExePath, err := tempbuild.Build(actualMainPkg, "dkt.")
+	if err != nil {
+		fmt.Fprintf(stderr, "ERROR: failed to build dkt: %v\n", err)
+		fmt.Fprintf(stderr, "--- diagnostics follow ---\n")
+		_ = printDiagnostics("", stderr)
+
+		return 1
+	}
+	if !keepExecutable {
+		defer os.Remove(dktExePath)
+	}
+
+	if debugTraceEnabled {
+		fmt.Fprintf(stderr, "%s%v\n", debugPrefix, append([]string{dktExePath}, args...))
+	}
+
+	return runDkt(stdin, stdout, stderr, dktExePath, args)
+}
+
+func runDkt(stdin io.Reader, stdout, stderr io.Writer, dktExePath string, args []string) int {
+	cmd := exec.Command(dktExePath, args...)
+
+	cmd.Stdin = stdin
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	signal.Ignore(os.Interrupt)
+
+	if err := cmd.Run(); err != nil {
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			return exitErr.ProcessState.ExitCode()
+		}
+
+		return 1
+	}
+
+	return 0
+}
+
+//------------------------------------------------------------------------------
+
+func printRunnerBuildInfo(prefix string, w io.Writer) {
+	buildInfo, ok := debug.ReadBuildInfo()
+	if !ok {
+		fmt.Fprintf(w, "%sdkt runner from GOPATH (%s)\n", prefix, runtime.Version())
+
+		return
+	}
+
+	mainMod := buildInfo.Main
+	fmt.Fprintf(w, "%sdkt runner %s (%s)\n", prefix, mainMod.Version, runtime.Version())
+}
+
+func printDiagnostics(prefix string, w io.Writer) error {
+	curMod, err := goListCurrentModule(prefix, w)
+	if err != nil {
+		return fmt.Errorf("goListCurrentModule: %w", err)
+	}
+
+	if curMod == "" {
+		fmt.Fprintf(w, "%snot in module-aware mode\n", prefix)
+	} else {
+		fmt.Fprintf(w, "%scurrent module: %q\n", prefix, curMod)
+	}
+
+	mainPkg, err := goListActualMainPkg(prefix, w)
+	if err != nil {
+		return fmt.Errorf("goListActualMainPkg: %w", err)
+	}
+
+	if mainPkg.Module == nil {
+		fmt.Fprintf(w, "%sfound dkt inside the GOPATH: %q\n", prefix, mainPkg.Dir)
+
+		return nil
+	}
+
+	mainMod := mainPkg.Module
+
+	var replaceText string
+	if replace := mainMod.Replace; replace != nil {
+		replaceText = fmt.Sprintf(" Replace={Path=%q Version=%q)", replace.Path, replace.Version)
+	}
+
+	fmt.Fprintf(w, "%sfound dkt module: Path=%q Version=%q%s\n",
+		prefix, mainMod.Path, mainMod.Version, replaceText)
+
+	return nil
+}
+
+type goListInfo struct {
+	Dir    string
+	Module *moduleInfo
+}
+
+type moduleInfo struct {
+	Path    string
+	Version string
+	Replace *moduleInfo
+	Main    bool
+	Dir     string
+}
+
+func goListCurrentModule(prefix string, w io.Writer) (string, error) {
```

Comment:
I'm not sure it would actually improve readability here or if it's just been on my mind lately but you could consider returning a `bool` to indicate whether this function (and also `goListActualMainPkg()`, though i think it would be slightly more awkward there) determined it was in module mode.

The callsite would be
```go
	curMod, inModuleMode, err := goListCurrentModule(prefix, w)
	if err != nil {
		return fmt.Errorf("goListCurrentModule: %w", err)
	}

	if !inModuleMode {
		fmt.Fprintf(w, "%snot in module-aware mode\n", prefix)
	} else {
		fmt.Fprintf(w, "%scurrent module: %q\n", prefix, curMod)
	}
```

Reply:
In this case, I don't think separating the module name and its existence makes the information clearer. A middle ground would be to return a `*string` so that you could return `nil` when not in module mode, but you'll never be in module mode and have a blank module name, so that seems overly complex.

Reply:
yeah we can certainly leave it as-is.